File: python/ch10_agent/agent_ch10/core/profiles.py
"""固定章节能力快照。

Java 对照：`ChapterProfile` 类似不可变配置 record，P01-P10 类似预定义单例常量。
组合根使用对象身份判断，调用方不能临时拼一个同字段对象来冒充正式章节。

这是什么：章节能力配置模块，定义每章可用的功能集合
为什么需要：
    1. 每章逐步引入新能力（loop → tools → policy → hooks...）
    2. Bootstrap 根据 Profile 决定装配哪些组件
    3. 单例模式防止伪造 Profile
"""

from dataclasses import dataclass
from typing import Literal

Capability = Literal[
    "loop",          # 基础 Agent 循环
    "powershell",    # PowerShell 工具
    "tool_registry", # 工具注册表
    "files",         # 文件读写工具
    "policy",        # 权限策略
    "hooks",         # 生命周期钩子
    "todo",          # TODO 管理
    "subagent",      # 子 Agent 调用
    "skills",        # 按需加载 Skill
    "artifacts",     # Artifact 生成
    "compaction",    # 历史压缩
    "memory",        # 长期记忆
    "dynamic_prompt",# 动态 System Prompt
]


@dataclass(frozen=True, slots=True)
class ChapterProfile:
    """章节号及其允许装配的能力白名单。

    这是什么：不可变配置对象，定义某章可用的功能集合
    Java 类比：不可变 record，类似 @ConfigurationProperties
    为什么需要：
        1. 明确每章的能力边界（逐章递增）
        2. Bootstrap 根据白名单决定装配哪些组件
        3. frozen=True 防止运行时修改

    字段：
        chapter: 章节号（1-20）
        capabilities: 允许的能力集合（frozenset 不可变）
    """

    chapter: int
    capabilities: frozenset[Capability]


P01 = ChapterProfile(1, frozenset({"loop", "powershell"}))
P02 = ChapterProfile(2, frozenset({"loop", "powershell", "tool_registry", "files"}))
P03 = ChapterProfile(3, frozenset({"loop", "powershell", "tool_registry", "files", "policy"}))
P04 = ChapterProfile(
    4, frozenset({"loop", "powershell", "tool_registry", "files", "policy", "hooks"})
)
P05 = ChapterProfile(
    5, frozenset({"loop", "powershell", "tool_registry", "files", "policy", "hooks", "todo"})
)
P06 = ChapterProfile(
    6,
    frozenset(
        {"loop", "powershell", "tool_registry", "files", "policy", "hooks", "todo", "subagent"}
    ),
)
P07 = ChapterProfile(
    7,
    frozenset(
        {
            "loop",
            "powershell",
            "tool_registry",
            "files",
            "policy",
            "hooks",
            "todo",
            "subagent",
            "skills",
        }
    ),
)

P08 = ChapterProfile(
    8,
    frozenset(
        {
            "loop",
            "powershell",
            "tool_registry",
            "files",
            "policy",
            "hooks",
            "todo",
            "subagent",
            "skills",
            "artifacts",
            "compaction",
        }
    ),
)

P09 = ChapterProfile(
    9,
    frozenset(
        {
            "loop",
            "powershell",
            "tool_registry",
            "files",
            "policy",
            "hooks",
            "todo",
            "subagent",
            "skills",
            "artifacts",
            "compaction",
            "memory",
        }
    ),
)

P10 = ChapterProfile(
    10,
    frozenset(
        {
            "loop",          # 第 1 章：基础循环
            "powershell",    # 第 1 章：PowerShell 工具
            "tool_registry", # 第 2 章：工具注册表
            "files",         # 第 2 章：文件读写
            "policy",        # 第 3 章：权限策略
            "hooks",         # 第 4 章：生命周期钩子
            "todo",          # 第 5 章：TODO 管理
            "subagent",      # 第 6 章：子 Agent
            "skills",        # 第 7 章：按需加载 Skill
            "artifacts",     # 第 8 章：Artifact 生成
            "compaction",    # 第 8 章：历史压缩
            "memory",        # 第 9 章：长期记忆
            "dynamic_prompt",# 第 10 章：动态 System Prompt（本章新增）
        }
    ),
)


def profile_for_chapter(chapter: int) -> ChapterProfile:
    """只返回模块内固定单例；尚未迁移的章节明确报错。

    这是什么：Profile 工厂方法，根据章节号返回对应的单例
    Java 类比：类似 enum 的 valueOf() 或单例注册表
    为什么需要：
        1. 集中管理所有 Profile，避免调用方直接引用 P01-P10
        2. 明确报错未迁移的章节（而不是返回 None）
        3. 返回单例保证对象身份（is 比较）

    参数：
        chapter: 章节号（1-20）

    返回：
        对应章节的 ChapterProfile 单例

    抛出：
        ValueError: chapter 不是 1-20 的整数，或该章尚未迁移
    """
    profiles = {
        1: P01,
        2: P02,
        3: P03,
        4: P04,
        5: P05,
        6: P06,
        7: P07,
        8: P08,
        9: P09,
        10: P10,
    }
    if isinstance(chapter, bool) or not isinstance(chapter, int) or not 1 <= chapter <= 20:
        raise ValueError("chapter 必须是 1 到 20 的整数")
    if chapter not in profiles:
        raise ValueError(f"第 {chapter} 章尚未迁移为 Python")
    return profiles[chapter]

File: typescript/chapters/ch17/src/core/profiles.ts
// 章节能力清单：每个 profile 都是冻结单例，组合根据此决定启用的运行时边界。
export type Capability =
  | "loop"
  | "powershell"
  | "tool_registry"
  | "files"
  | "policy"
  | "hooks"
  | "todo"
  | "subagent"
  | "skills"
  | "artifacts"
  | "compaction"
  | "memory"
  | "dynamic_prompt"
  | "recovery"
  | "task_dag_json"
  | "background"
  | "cron"
  | "teammate"
  | "mailbox"
  // protocol 与 plan_gate 共同表示结构化审批消息和副作用执行门控。
  | "protocol"
  | "plan_gate"
  // task_dag_sqlite 与 work_stealing 表示带租约的 SQLite 任务图与去中心化认领。
  | "task_dag_sqlite"
  | "work_stealing";

export interface ChapterProfile {
  // chapter 与能力集合共同构成固定快照，bootstrap 通过引用身份拒绝伪造 profile。
  readonly chapter: number;
  readonly capabilities: ReadonlySet<Capability>;
}

class CapabilitySet implements ReadonlySet<Capability> {
  // 内部 Set 不向外暴露，所有迭代方法只提供只读视图。
  readonly #values: Set<Capability>;

  constructor(values: readonly Capability[]) {
    // 复制后冻结输入语义，调用方后续修改原数组不会改变章节能力。
    this.#values = new Set(values);
  }

  get size(): number {
    return this.#values.size;
  }

  has(value: Capability): boolean {
    // 能力判断集中委托内部 Set，组合根不依赖数组顺序。
    return this.#values.has(value);
  }

  entries(): SetIterator<[Capability, Capability]> {
    // ReadonlySet 迭代协议保持与原生 Set 一致。
    return this.#values.entries();
  }

  keys(): SetIterator<Capability> {
    // key/value 对于集合相同，均返回内部只读迭代器。
    return this.#values.keys();
  }

  values(): SetIterator<Capability> {
    // 返回迭代器而非底层 Set，防止调用方获得可变集合引用。
    return this.#values.values();
  }

  forEach(
    callbackfn: (value: Capability, value2: Capability, set: ReadonlySet<Capability>) => void,
    thisArg?: unknown,
  ): void {
    // 回调第三参数仍传当前只读包装，不能通过它修改内部 Set。
    this.#values.forEach((value) => {
      callbackfn.call(thisArg, value, value, this);
    });
  }

  [Symbol.iterator](): SetIterator<Capability> {
    return this.values();
  }
}

export const P01: ChapterProfile = Object.freeze({
  chapter: 1,
  capabilities: new CapabilitySet(["loop", "powershell"]),
});

export const P02: ChapterProfile = Object.freeze({
  chapter: 2,
  capabilities: new CapabilitySet(["loop", "powershell", "tool_registry", "files"]),
});

export const P03: ChapterProfile = Object.freeze({
  chapter: 3,
  capabilities: new CapabilitySet(["loop", "powershell", "tool_registry", "files", "policy"]),
});

export const P04: ChapterProfile = Object.freeze({
  chapter: 4,
  capabilities: new CapabilitySet([
    "loop",
    "powershell",
    "tool_registry",
    "files",
    "policy",
    "hooks",
  ]),
});

export const P05: ChapterProfile = Object.freeze({
  chapter: 5,
  capabilities: new CapabilitySet([
    "loop",
    "powershell",
    "tool_registry",
    "files",
    "policy",
    "hooks",
    "todo",
  ]),
});

export const P06: ChapterProfile = Object.freeze({
  chapter: 6,
  capabilities: new CapabilitySet([
    "loop",
    "powershell",
    "tool_registry",
    "files",
    "policy",
    "hooks",
    "todo",
    "subagent",
  ]),
});

export const P07: ChapterProfile = Object.freeze({
  chapter: 7,
  capabilities: new CapabilitySet([
    "loop",
    "powershell",
    "tool_registry",
    "files",
    "policy",
    "hooks",
    "todo",
    "subagent",
    "skills",
  ]),
});

export const P08: ChapterProfile = Object.freeze({
  chapter: 8,
  capabilities: new CapabilitySet([
    "loop",
    "powershell",
    "tool_registry",
    "files",
    "policy",
    "hooks",
    "todo",
    "subagent",
    "skills",
    "artifacts",
    "compaction",
  ]),
});

export const P09: ChapterProfile = Object.freeze({
  chapter: 9,
  capabilities: new CapabilitySet([
    "loop",
    "powershell",
    "tool_registry",
    "files",
    "policy",
    "hooks",
    "todo",
    "subagent",
    "skills",
    "artifacts",
    "compaction",
    "memory",
  ]),
});

export const P10: ChapterProfile = Object.freeze({
  chapter: 10,
  capabilities: new CapabilitySet([
    "loop",
    "powershell",
    "tool_registry",
    "files",
    "policy",
    "hooks",
    "todo",
    "subagent",
    "skills",
    "artifacts",
    "compaction",
    "memory",
    "dynamic_prompt",
  ]),
});

export const P11: ChapterProfile = Object.freeze({
  chapter: 11,
  capabilities: new CapabilitySet([
    "loop",
    "powershell",
    "tool_registry",
    "files",
    "policy",
    "hooks",
    "todo",
    "subagent",
    "skills",
    "artifacts",
    "compaction",
    "memory",
    "dynamic_prompt",
    "recovery",
  ]),
});

export const P12: ChapterProfile = Object.freeze({
  chapter: 12,
  capabilities: new CapabilitySet([...P11.capabilities, "task_dag_json"]),
});

export const P13: ChapterProfile = Object.freeze({
  chapter: 13,
  capabilities: new CapabilitySet([...P12.capabilities, "background"]),
});

export const P14: ChapterProfile = Object.freeze({
  chapter: 14,
  capabilities: new CapabilitySet([...P13.capabilities, "cron"]),
});

export const P15: ChapterProfile = Object.freeze({
  chapter: 15,
  capabilities: new CapabilitySet([...P14.capabilities, "teammate", "mailbox"]),
});

export const P16: ChapterProfile = Object.freeze({
  chapter: 16,
  capabilities: new CapabilitySet([...P15.capabilities, "protocol", "plan_gate"]),
});

// P17 在 P16 基础上追加 SQLite 任务图与 work stealing，让 Lead、子代理和队友共享同一认领路径。
export const P17: ChapterProfile = Object.freeze({
  chapter: 17,
  capabilities: new CapabilitySet([...P16.capabilities, "task_dag_sqlite", "work_stealing"]),
});

export function profileForChapter(chapter: number): ChapterProfile {
  if (chapter === 1) {
    return P01;
  }
  if (chapter === 2) {
    return P02;
  }
  if (chapter === 3) {
    return P03;
  }
  if (chapter === 4) {
    return P04;
  }
  if (chapter === 5) {
    return P05;
  }
  if (chapter === 6) {
    return P06;
  }
  if (chapter === 7) {
    return P07;
  }
  if (chapter === 8) {
    return P08;
  }
  if (chapter === 9) {
    return P09;
  }
  if (chapter === 10) {
    return P10;
  }
  if (chapter === 11) {
    return P11;
  }
  if (chapter === 12) {
    return P12;
  }
  if (chapter === 13) {
    return P13;
  }
  if (chapter === 14) {
    return P14;
  }
  if (chapter === 15) {
    return P15;
  }
  if (chapter === 16) {
    return P16;
  }
  if (chapter === 17) {
    return P17;
  }
  if (!Number.isInteger(chapter) || chapter < 1 || chapter > 20) {
    throw new Error("chapter must be an integer from 1 to 20");
  }
  throw new Error(`Chapter ${chapter} has not been migrated to TypeScript yet`);
}
